src/Interpreters/AnyInputOptimize.cpp

```diff
@@ -0,0 +1,81 @@
+#include <Common/typeid_cast.h>
+#include <Parsers/ASTLiteral.h>
+#include <Parsers/ASTFunction.h>
+#include <Parsers/ASTIdentifier.h>
+#include <Interpreters/AnyInputOptimize.h>
+#include <AggregateFunctions/AggregateFunctionFactory.h>
+#include <IO/WriteHelpers.h>
+#include <Parsers/ASTTablesInSelectQuery.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+    extern const int UNEXPECTED_AST_STRUCTURE;
+}
+
+namespace
+{
+    constexpr const char * any = "any";
+}
+
+ASTPtr * getExactChild(const ASTPtr & ast, const size_t ind)
+{
+    return &ast->as<ASTFunction>()->arguments->children[ind];
+}
+
+///recursive searching of identifiers
+void changeAllIdentifiers(ASTPtr & ast, size_t ind)
+{
+    if ((*getExactChild(ast, ind))->as<ASTIdentifier>())
+    {
+        ///put new any
+        ASTPtr oldAST = (*getExactChild(ast, ind));
+        (*getExactChild(ast, ind)) = makeASTFunction("any");
+        (*getExactChild(ast, ind))->as<ASTFunction>()->arguments->children.push_back(oldAST);
+    }
+    else if ((*getExactChild(ast, ind))->as<ASTFunction>() &&
+        !AggregateFunctionFactory::instance().isAggregateFunctionName((*getExactChild(ast, ind))->as<ASTFunction>()->name))
+        for (size_t i = 0; i < (*getExactChild(ast, ind))->as<ASTFunction>()->arguments->children.size(); i++)
+            changeAllIdentifiers((*getExactChild(ast, ind)), i);
+    else if (ast->as<ASTFunction>()->arguments->children[ind]->as<ASTFunction>() &&
+             AggregateFunctionFactory::instance().isAggregateFunctionName((*getExactChild(ast, ind))->as<ASTFunction>()->name))
+        throw Exception("Aggregate function is found inside another aggregate function in query", 184);
+}
+
+
+///cut old any, put any to identifiers. any(functions(x)) -> functions(any(x1))
+void AnyInputMatcher::visit(ASTPtr & current_ast, Data data)
+{
+    data = {};
+    if (!current_ast)
+        return;
+
+    auto * function_node = current_ast->as<ASTFunction>();
+    if (function_node && function_node->name == "any" && function_node->arguments->children[0]->as<ASTFunction>())
+    {
+        ///cut any
+        current_ast = function_node->arguments->children[0];
```

Comment:
You have to clone ast here, otherwise it won't work with aliases.
Consider query:
`SELECT any(number * 2) as n, n * 2 FROM numbers(100)`
At the optimization stage alias will be replaced with actual expression, and after you will modify `number * 2` to `any(number) * 2` while optimizing first expression, the second will look like `any(any(number) * 2)`